tsweb/jsonhandler.go

```diff
@@ -0,0 +1,280 @@
+	//
+	// The caller will write a response to the ResponseWriter.
+	// The response is always a JSON blob, either:
+	//
+	//	{"success": true, "data": <encoded obj>}
+	//
+	// or
+	//
+	//	{"success": false, "error": "<error message>"}
+	//
+	// If ServeHTTPJSON returns no error, the returned obj is
+	// encoded as a JSON property list and sent to the client.
+	//
+	// If ServeHTTPJSON returns an error, it caller should handle
+	// an error by serving an HTTP 500 response to the user. The
+	// error details must not be sent to the client, as they may
+	// contain sensitive information. If the error is an
+	// HTTPError, though, callers should use the HTTP response
+	// code and message as the response to the client.
+	ServeHTTPJSON(http.ResponseWriter, *http.Request, map[string]interface{}) (obj interface{}, err error)
+}
+
+// StdJSONHandler converts a JSONHandler into a standard http.Handler.
+// Handled requests are logged using logf, as are any errors. Errors
+// are handled as specified by the Handler interface.
+func StdJSONHandler(jh JSONHandler, logf logger.Logf) http.Handler {
+	return jsonHandler{jh: jh, logf: logf, timeNow: time.Now, log200s: true}
+}
+
+// JSONHandlerFunc is an adapter to allow the use of ordinary
+// functions as JSONHandlers. If f is a function with the
+// appropriate signature, JSONHandlerFunc(f) is a JSONHandler that
+// calls f.
+//
+// See the documentation on JSONHandler.ServeHTTPJSON for semantics
+// of the parameters.
+type JSONHandlerFunc func(http.ResponseWriter, *http.Request, map[string]interface{}) (interface{}, error)
+
+// ServeHTTPJSON calls f(w, r).
+func (f JSONHandlerFunc) ServeHTTPJSON(w http.ResponseWriter, r *http.Request, data map[string]interface{}) (interface{}, error) {
+	return f(w, r, data)
+}
+
+// jsonHandler is an http.Handler that wraps a Handler and handles JSON.
+type jsonHandler struct {
+	jh      JSONHandler
+	logf    logger.Logf
+	timeNow func() time.Time
+	log200s bool
+}
+
+func parseFormData(w http.ResponseWriter, r *http.Request) (map[string]interface{}, error) {
+	postData := make(map[string]interface{})
+	ct, _, _ := mime.ParseMediaType(r.Header.Get("Content-Type"))
+	switch ct {
+	case "application/x-www-form-urlencoded":
+		if err := r.ParseForm(); err != nil {
+			return nil, err
+		}
+		for k, values := range r.Form {
+			if len(values) == 1 {
+				postData[k] = values[0]
+			} else {
+				postData[k] = values
+			}
+		}
+	default:
+		r.Body = http.MaxBytesReader(w, r.Body, 1<<20)
+		b, err := ioutil.ReadAll(r.Body)
+		r.Body.Close()
+		if err != nil {
+			return nil, err
+		}
+		if len(b) == 0 {
+			return nil, err
+		}
+		if err := json.Unmarshal(b, &postData); err != nil {
+			return nil, fmt.Errorf("json decode request: %w", err)
+		}
+	}
+	return postData, nil
+}
+
+// ServeHTTP implements the http.Handler interface.
+func (h jsonHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	msg := AccessLogRecord{
```

Comment:
seems like we shouldn't need to re-implement all this logging stuff if we compose the existing pieces in a different way